### PR TITLE
From subtest to the official first release of the mplc package

### PR DESCRIPTION
**Pull request by arthurPignet**
The refacto to a pypi package has led to deep changes in the project architecture, and the documentation needs to be updated.

### CODE REVIEW

File: subtest/docs/documentation.md [hunk 1]
```diff
@@ -162,6 +162,159 @@ print(contributivity_score[0])
...
+
+```python
+ scenario_params['dataset_name']=['mnist', 'cifar10']
+```
+
+
+
+Please, note that if you want to specify a path to saved weights and launch the same scenario with a random initialization, you must specify it in the list of initialization.
+
+```
```
Comment (natct10) [inline subtest/docs/documentation.md:181]:
```suggestion
```yaml
```
You can add the language here

File: subtest/docs/pck-description.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+       epoch_count:
+         - 38
+       methods:
+         - ["Shapley values", "Independent scores", "TMCS"]
+       minibatch_count:
+         - 20
+       gradient_updates_per_pass_count:
+         - 8
+       dataset_proportion:
+	     - 1
```
Comment (natct10) [inline subtest/docs/pck-description.md:119]: indentation pb here

File: subtest/docs/pck-description.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+       epoch_count:
+         - 38
+       methods:
+         - ["Shapley values", "SMCS", "IS_lin_S", "IS_reg_S"]
+       minibatch_count:
+         - 20
+       gradient_updates_per_pass_count:
+         - 8
+       dataset_proportion:
+	     - 1
```
Comment (natct10) [inline subtest/docs/pck-description.md:143]: indentation pb here

**Review by natct10 — CHANGES_REQUESTED** [review]
Looks nice, thx @arthurPignet !
Could you please add the language markup when using code block ("```python")

File: README.md [hunk 4]
```diff
@@ -5,27 +5,19 @@
 # Exploration of dataset contributivity to a model in collaborative ML projects
 
 - [Introduction](#introduction)
-  - [Context of this work](#context-of-this-work)
-  - [How to interact with this?](#how-to-interact-with-this)
+  * [Context of this work](#context-of-this-work)
+  * [How to interact with this?](#how-to-interact-with-this-)
```
Comment (natct10) [inline README.md:9]: the link here seems to be broken?

File: subtest/docs/documentation.md [hunk 5]
```diff
@@ -2,111 +2,85 @@
...
-
-#### Dataset
-
-This is where you define your dataset and relatives objects such as preprocessing functions and model generator.
-
-##### Dataset
-
-This is the structure of the dataset generator:
+### My first scenario
+To launch a collaborative round, you need to generate scenario and run it, though a Scenario object.
```
Comment (natct10) [inline subtest/docs/documentation.md:39]:
```suggestion
To launch a collaborative round, you need to generate scenario and run it, though a `Scenario` object.
```

File: subtest/docs/documentation.md [hunk 6]
```diff
@@ -2,111 +2,85 @@
...
-    input_shape,
-    num_classes,
-    preprocess_dataset_labels,      # See below
-    generate_new_model_for_dataset  # See below
-)
+from subtest.scenario import Scenario
+my_scenario = Scenario(partners_count=3,
+                       amounts_per_partner=[0.2, 0.3, 0.5])
 ```
+Note that you can use more advanced sample split options in order to fine tune the data distribution between partners. See the doc
```
Comment (natct10) [inline subtest/docs/documentation.md:49]: It could be nice to point towards a specific section on "set the documentation".

File: subtest/docs/documentation.md [hunk 7]
```diff
@@ -145,22 +120,225 @@ sns.relplot(data = df, kind = "line")
...
 ```
 
-Check out our Tutorial 3 for more information.
+Check out our [Tutorial 3](https://github.com/SubstraFoundation/distributed-learning-contributivity/blob/master/notebooks/examples/3_Exploring_results.ipynb) for more information.
 
-## Contributivity measurement methods
+### Contributivity measurement methods
 
-To use contributivity measurement tools, you will have to set parameters to your `scenario` object.
+To use contributivity measurement tools, you will have to change the parameters of your Scenario object
```
Comment (natct10) [inline subtest/docs/documentation.md:127]:
```suggestion
To use contributivity measurement tools, you will have to change the parameters of your `Scenario` object
```

File: subtest/docs/pck-description.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+### Experimental approach
+
+We want to start experimenting contributivity evaluations in collaborative data science and distributed multi-partner learning scenarios. Our exploration of this topic is in progress, as is this library and associated experiments. To make the most out of it, it is key to capitalize on this effort and develop it as a reproducible pipeline that can be improved, enriched, complemented over time.
+
+For a start we made the following choices:
+
+- What we want to compare (with the Shapley values being the baseline, see section below):
+  - Contributivity relative values
+  - Computation time
+- Public datasets for experiments currently supported: MNIST, CIFAR10, TITANIC
```
Comment (natct10) [inline subtest/docs/pck-description.md:64]:
```suggestion
- Public datasets for experiments currently supported: MNIST, CIFAR10, TITANIC and ESC50
```

File: subtest/docs/pck-description.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+``` 
+scenario_params_list:
+ - dataset_name:
+    - 'mnist'
+    - 'cifar10'
+```
+
+Please, note that if you want to specify a path to saved weights and launch the same scenario with a random initialization, you must specify it in the list of initialization.
+
+```
```
Comment (natct10) [inline subtest/docs/pck-description.md:192]:
```suggestion
```yaml
```

**Review by natct10 — CHANGES_REQUESTED** [review]
This is a great work you have done here, thank you @arthurPignet!
I have suggested some edits and I'm likely to add others on the markdown formatting, but this is already great and the examples you added for params are awesome!

**Review by bowni — COMMENTED** [review]
For the refacto complements it all looks good on my end - I didn't test though, I trust @arthurPignet that you ran tests yourself :crossed_fingers: 
For the documentation part, I am on it now :smiley:

File: main.py [hunk 10]
```diff
@@ -70,7 +70,7 @@ def main():
                     **scenario_params,
                     experiment_path=experiment_path,
                     scenario_id=scenario_id + 1,
-                    n_repeat=i + 1
+                    repeat_tag=i + 1
```
[the remark this answers is not in the document]
Comment (arthurPignet) [reply main.py:86]: Hm in fact I changed it in scenario, because the parameter n_repeat (at Scenario object level), is only used for the save-name of the scenario. But in main, maybe n_repeat is better

File: README.md [hunk 11]
```diff
@@ -83,7 +75,9 @@ Once a given scenario configured, it seemed useful to choose how the multi-partn
 
 Finally, with given scenarios and multi-partner learning approaches, we can address contributivity measurement approaches. See below sections [Configuration of contributivity measurement methods to be tested](#configuration-of-contributivity-measurement-methods-to-be-tested) and [Contributivity measurement approaches studied and implemented](#contributivity-measurement-approaches-studied-and-implemented).
 
-### Using the code files
+### Run an experiment
+
+#### Using the code files
```
Comment (bowni) [inline README.md:80]: Maybe here, inbetween, a short text explaining that the intended usage is as a library etc., but it is also possible to run it the old fashion way etc.

File: subtest/docs/pck-description.md [hunk 12]
```diff
@@ -0,0 +1,405 @@
+[![Build Status](https://travis-ci.org/SubstraFoundation/distributed-learning-contributivity.svg?branch=master)](https://travis-ci.org/SubstraFoundation/distributed-learning-contributivity)
```
Comment (bowni) [inline subtest/docs/pck-description.md:1]: Ok so if I understand correctly the entire old README content is now in this `pck-description.md` file. What is the rationale for this? As you kept some bits in the old README, it means we'll need to maintain duplicates of the same texts?
Maybe I just wonder about that because I am not familiar with the use of this `pck-description.md` file :smiley:

File: subtest/docs/documentation.md [hunk 13]
```diff
@@ -2,111 +2,85 @@
...
+  * [Train/validation/test splits](#train-validation-test-splits)
+- [Contacts, contributions, collaborations](#contacts--contributions--collaborations)
+
+## Prerequisites
 
 At root folder:
 ```bash
 pip install -r requirements.txt
-pip install -i https://test.pypi.org/simple/ subtest==0.0.0.6
+pip install -i https://test.pypi.org/simple/ subtest==0.0.0.18
```
Comment (bowni) [inline subtest/docs/documentation.md:31]: Let's rename the package to `mplc-simulations`, it is time... :smiley:

**Review by bowni — COMMENTED** [review]
It is too cumbersome to review text files with that many changes. I'll move to my IDE and commit changes directly

File: subtest/docs/pck-description.md (hunk 12 above)
Comment (arthurPignet) [reply subtest/docs/pck-description.md:1]: Initially the pck_description was the old README, but as the README has changed, I keep the old description in this pck-description. It is maybe not really useful

**Comment by bowni** [comment]
@arthurPignet is it normal to still have a `/shapley_value` folder at the root level since your last refactos?

**Comment by arthurPignet** [comment]
> @arthurPignet is it normal to still have a `/shapley_value` folder at the root level since your last refactos?

No, indeed it is useless, unless if we want to keep the full repo which is cloned

**Comment by bowni** [comment]
Ok, I pushed just a few amendments to the texts - there is still more to do. I can do it next week (and/or @arthurPignet you can see the style and tonality and give it a go too, as you want). It will be a collective effort anyway!

File: mplc/docs/pck-description.md [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+
+#### Contributivity measurement approaches
+
+Finally, with given scenarios and multi-partner learning approaches, we can address contributivity measurement approaches. See below sections [Configuration of contributivity measurement methods to be tested](#configuration-of-contributivity-measurement-methods-to-be-tested) and [Contributivity measurement approaches studied and implemented](#contributivity-measurement-approaches-studied-and-implemented).
+
+### Using the code files
+
+1. Define your mock scenario(s) in `config.yml` by changing the values of the suggested parameters of the 2 example scenarios (you can browse more available parameters in section [Config file parameters](#config-file-parameters) below). For example:
+
+    ```yaml
```
Comment (natct10) [inline mplc/docs/pck-description.md:90]: indentation

File: mplc/docs/documentation.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+`partners_count`: `int`  
+Number of partners in the mocked collaborative ML scenario.  
+Example: `partners_count: 4`
+
+`amounts_per_partner`: `[float]`  
+Fractions of the original dataset each partner receives to mock a collaborative ML scenario where each partner provides data for the ML training.  
+You have to ensure the fractions sum up to 1.
+Example: `amounts_per_partner: [0.3, 0.3, 0.1, 0.3]`
+
+<a id="sample_split_option"></a>
```
Comment (natct10) [inline mplc/docs/documentation.md:205]: Is it normal?

File: mplc/docs/documentation.md [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    # add layers
+    model.add(Dense(num_classes, activation='softmax'))
+    # compile with loss and accuracy
+    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
+    return model
+```
+Note: It is mandatory to have loss and accuracy as metrics for your model.
+
+### Preprocessing
+```python
```
Comment (natct10) [inline mplc/docs/documentation.md:349]: line break

**Review by natct10 — CHANGES_REQUESTED** [review]
This is a great step @arthurPignet, well done!

File: mplc/docs/documentation.md (hunk 15 above)
Comment (natct10) [reply mplc/docs/documentation.md:205]: => anchor

**Review by bowni — APPROVED** [review]

**Comment by bowni** [comment]
Just 1 detail I noticed: in the tutorial the logs says subtest and not mplc:
`2020-09-18 11:01:31.280 | DEBUG    | subtest.scenario:__init__:118 - Dataset selected: mnist`

Let's keep that in mind for a next improvement iteration!

**Review by natct10 — APPROVED** [review]
Well done @arthurPignet, this is a really nice job you have done here!
Thank you for this nice contribution!

File: mplc/utils.py [hunk 17]
```diff
@@ -126,3 +128,73 @@ def init_result_folder(yaml_filepath, cfg):
...
+    def set_to_info_level(self):
+        self.level = "INFO"
+
+    def __call__(self, record):
+        levelno = logger.level(self.level).no
+        return record["level"].no >= levelno
+
+
+def set_log_file(path):
+    logger.remove()
```
Comment (RomainGoussault) [inline mplc/utils.py:195]: Why do you need this line here ?
I will remove the handler added on line 171

File: mplc/scenario.py [hunk 18]
```diff
@@ -43,10 +43,40 @@ def __init__(self,
...
+        :param aggregation_weighting: 'uniform' (default), 'data_volume' or 'local_score'
+        :param gradient_updates_per_pass_count: int
+        :param minibatch_count: int
+        :param epoch_count: int
+        :param is_early_stopping: boolean. Stop the training if scores on val_set reach a plateau
+        :param methods: A declarative list `[]` of the contributivity measurement methods to be executed.
+        :param is_quick_demo: boolean. Useful for debugging
+        :param experiment_path: path
+        :param scenario_id: str
+        :param repeat_tag: int
```
Comment (RomainGoussault) [inline mplc/scenario.py:76]: I think repeat tag instead of n_repeats ? or repeats_count ? 
The `tag` makes me feel it's a flag, or boolean but not an actual number.

**Review by RomainGoussault — APPROVED** [review]
Terrific @arthurPignet !!!

I have made a few comments, but these are very minor issues.
On the notebooks, I have noted two things to be addressed:

- On the first notebook, you have set the xlabel and ylabel but somethow it does not show on the graph
- On notebook 3 you have an error on the last cell

As I said, these are minor comments, feel free to merge and adress these issues later.

File: mplc/utils.py (hunk 17 above)
Comment (arthurPignet) [reply mplc/utils.py:195]: If you call this function twice, for instance when you change the folder for a new scenario, you will add the new log files. Not replace the old ones. 
So at the end, you will write on every file you set up. 
With this line, I remove every handler, (and I instantly put back the default handler) 
An other solution would have been saving the handler id, and removing it. But where could be stored the id, since we are on a library context ? 
So I choose to remove all handlers. 
At the ends, since a library is not really supposed to log anything, I make something the simplest as I could, which matches the main.py.

File: mplc/scenario.py (hunk 18 above)
Comment (arthurPignet) [reply mplc/scenario.py:76]: ok I see. I wanted to avoid a confusion here, as the n_repeat can be easily interpreted as number of repeat. However, setting this parameter (at scenario level) to 5 only means that the folder will be numerated 5, not that the scenario will run 5 times.  
I go for repeat number.  
Maybe this can be changed by adding a name to a scenario, what do you think ? For the future

Comment (arthurPignet) [reply mplc/scenario.py:76]: sorry, repeat_count, not number
